### DISCUSSION TITLE
Add edge bundling algorithms

### DISCUSSION THREAD

**Discussion by ianthomas23** [networkx/networkx#6140] [Algorithms]
Hello, Datashader maintainer here!

The idea of moving some of the Datashader edge bundling algorithms across to NetworkX has been aired, so I am creating this discussion so that we can work out how we might approach this.

The Datashader edge bundling algorithm has two stages:

1. Take the node (x, y) positions and edge connectivity, along with optional extra node- and edge-based data (e.g. weights), and calculate curves for each edge which are really piecewise linear paths. There may also be some extraction or interpolation of node- or edge-based data to the path coordinates.
2. Combine the lines in the usual Datashader manner by aggregating on a per-pixel basis and applying some (usually highly nonlinear) transform to shade the pixels using a colormap.

Stage 2 is core Datashader work, it is what the library exists to do. But stage 1 is arguably out of scope for Datashader and would better reside in a domain-specific graph processing library, e.g. NetworkX.

The existing code (https://github.com/holoviz/datashader/blob/master/datashader/bundling.py) uses Numba, which is common in Datashader, and is therefore quite low-level explicit-loop Python. I know that we cannot add Numba as a NetworkX dependency, so this code needs to be altered to be more NumPy-like, i.e. higher-level vectorised. But it is an acceptable starting point because with the Numba decorators removed it will run as pure Python code, albeit somewhat slowly. It will also be useful for comparing accuracy of the NumPy vs Numba implementations.

I think probably we have to get some code written as a demonstrator so that people can play around with a bit and only then might we know exactly how and where it might fit into NetworkX. I can see the possibility that it will useful in itself but of limited performance, so OK for small problems but for larger problems more powerful use of Numba/Dask/CUDA might be desirable, and then we would have to work out where that code would sit. But those sorts of decisions can be put off for the moment.

Maybe we should approach the problem from two directions, top-down and bottom-up:

1. Top-down = API. Incoming data will presumably be a form consistent with existing NetworkX usage. Will output format be similar? It will still be graph, just with curved edges. We should also consider that although we are starting with a single algorithm we may well want to add more in the future, so the API should support this. I am hoping that some NetworkX maintainers will take the lead with this as you won't want a NetworkX novice (me) making these decisions!
2. Bottom-up = the low-level code that calculates the curved line segments, hopefully implemented in a NumPy-like vectorised manner. I am expecting to start on this, but I will happily accept the help of anyone else who is interested. I am not an expert on these algorithms, but if someone on the Datashader team needs to be an expert than I would be that person.

**Comment by rossbar** [comment]
Thanks for starting the discussion @ianthomas23 !

This does seem interesting - your breakdown of the problem into the edge-bundling & and visualization components makes a lot of sense. I'm not an expert in networks so my opinion shouldn't be weighted heavily, but at face-value it does seem to me that step 1 of the edge-bundling procedure is a good fit for networkx as it appears to be generally applicable to networks. My gut feeling is also that it would be good to keep the actual *visualization* of the edge bundles in dedicated graph visualization packages (e.g. datashader), though we could potentially add some basic visualization to the `nx_pylab` subpackage in NetworkX itself.

I can offer my two cents on a potential API. One idea for the output could be a dictionary that maps edges (as 2-tuples, or 3-tuples if the edge bundling supports multiedges) to the piecewise linear paths. This is analogous to the pattern we use for node positions in the layout functions, which return dictionaries that map each node to a numpy array representing the coordinates of the center of the node. So the usage might looks something like:

```python
>>> import networkx as nx
>>> G = nx.Graph([(0, 1), (1, 2)])  # a path graph with 2 edges
>>> nx.bundle_edges(G, **whatever_kwargs_you_want)
{(0, 1): <path_obj_1>, (1, 2): <path_obj_2>}
```

As for the type of the `<path_obj>`s above, IMO it would be great if they were represented as numpy arrays, but maybe something else makes more sense.

**Comment by ianthomas23** [comment]
Thanks @rossbar. We will definitely keep the primary visualisation in Datashader as the edge rendering is just a slight specialisation of rendering a sequence of lines, so that is easily achieved. It should also be pretty easy for users to extract the edge coordinates and pass them to general rendering libraries like Bokeh or Matplotlib for rendering, regardless of whether there is specific visualisation of them NetworkX.

Your API idea sounds fine to me. We might also eventually need to consider if other consumers of the bundled edges prefer different formats such as dense NumPy arrays or Pandas Dataframes. I think the low-level code is likely to produce a single 2D NumPy array for the edge coordinates, which will compatible with any of these possibilities.

I will start looking at NumPy-fying the original algorithm and report back here.
